### PR TITLE
JAMES-2046 get message list SentDate sort should default to internalDate

### PR DESCRIPTION
**Pull request by chibenwa**
My rework of https://github.com/linagora/james-project/pull/832/

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/CombinedComparator.java [hunk 1]
```diff
@@ -95,14 +95,11 @@ private CombinedComparator(List<Comparator<MailboxMessage>> comparators) {
...
-            if (i != 0) {
-                break;
-            }
-        }
-        return i;
+        return comparators.stream()
+            .map(comparator -> comparator.compare(o1, o2))
+            .filter(result -> result != 0)
+            .findFirst()
+            .orElse(0);
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/CombinedComparator.java]: :eyeglasses:

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java [hunk 2]
```diff
@@ -18,44 +18,69 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(SentDateComparator.class);
+    // Some sent e-mail have this form : Wed,  3 Jun 2015 09:05:46 +0000 (UTC)
+    // Java 8 Time library RFC_1123_DATE_TIME corresponds to Wed,  3 Jun 2015 09:05:46 +0000 only
+    // This REGEXP is here to match ( in order to remove ) the possible invalid end of a header date
+    // Example of matching patterns :
+    //  (UTC)
+    //  (CEST)
+    private static final Pattern DATE_SANITIZING_PATTERN = Pattern.compile(" *\\(.*\\) *");
+
+    public static Optional<ZonedDateTime> toISODate(String value) {
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java]: Do will always know the timezone ? Or do we always use UTF as a zone ? If yes I would prefer `Instant`, we already have a debate on that

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java [hunk 3]
```diff
@@ -18,44 +18,69 @@
...
+            return Optional.of(ZonedDateTime.parse(
+                sanitizeDateStringHeaderValue(value),
+                ImapDateTimeFormatter.rfc5322()));
+        } catch (Exception e) {
+            LOGGER.info("Can not parse receive date " + value);
+            return Optional.empty();
+        }
+    }
+
+    @VisibleForTesting
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java]: Indent issue

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java [hunk 4]
```diff
@@ -18,44 +18,69 @@
...
+
     public final static Comparator<MailboxMessage> SENTDATE = new SentDateComparator();
+    private static final Logger LOGGER = LoggerFactory.getLogger(SentDateComparator.class);
+    // Some sent e-mail have this form : Wed,  3 Jun 2015 09:05:46 +0000 (UTC)
+    // Java 8 Time library RFC_1123_DATE_TIME corresponds to Wed,  3 Jun 2015 09:05:46 +0000 only
+    // This REGEXP is here to match ( in order to remove ) the possible invalid end of a header date
+    // Example of matching patterns :
+    //  (UTC)
+    //  (CEST)
+    private static final Pattern DATE_SANITIZING_PATTERN = Pattern.compile(" *\\(.*\\) *");
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java]: If this is neither (UTC) or (CEST) what should we do ignore it or fail ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/comparator/SentDateComparatorTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    public void sanitizeDateStringHeaderValueShouldRemoveCESTPart() {
+        assertThat(
+            SentDateComparator.sanitizeDateStringHeaderValue("Thu, 18 Jun 2015 04:09:35 +0200 (CEST)"))
+            .isEqualTo("Thu, 18 Jun 2015 04:09:35 +0200");
+    }
+
+    @Test
+    public void sanitizeDateStringHeaderValueShouldRemoveUTCPart() {
+        assertThat(
+            SentDateComparator.sanitizeDateStringHeaderValue("Thu, 18 Jun 2015 04:09:35 +0200  (UTC)  "))
```
Comment (strokyl) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/comparator/SentDateComparatorTest.java]: Can you dd a test to specify behaviour for a invalid date format 'example (GMT)'

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 6]
```diff
@@ -1143,4 +1146,30 @@ public void searchShouldReturnEmptyWhenMailboxBelongingToAnotherMailboxSession()
 
         assertThat(messageSearchIndex.search(otherSession, mailbox, searchQuery)).isEmpty();
     }
+
+    @Test
+    public void searchShouldOrderByInternalDateWhenSortOnSentDateAndNoCorrespondingHeader() throws Exception {
```
Comment (strokyl) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: I think my test name was better "sentDateComparatorTestShouldCompareInternalDateAgainstSentDateWhenOneMailOverTwoHasNoSentDate"

Comment (strokyl) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: Moreover I think one test is not enough we need three test:
 * none of the message has the header
 * one of them have it
 * all of them have it

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/CombinedComparator.java (hunk 1 above)
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/CombinedComparator.java]: Viva Java8

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java (hunk 2 above)
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java]: That code is moved from ES to avoid circular dependency and share same code path.

Maybe this code is orthogonal with the issue.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java (hunk 4 above)
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java]: That code is moved from ES to avoid circular dependency and share same code path.

Maybe this code is orthogonal with the issue.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java (hunk 6 above)
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: I don't agree, we are searching. Then it should be the first verb.

In ES there is no sentDateComparator. It is then pointless.

We already have a test with All SendDate. We just miss the Mixed test.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java (hunk 4 above)
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java]: This code just remove brackets and their content. It's duplicated from timezone offset and makes java8 parser crash.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java (hunk 2 above)
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java]: My answer would be : from the RFC there, we always now it.

**Comment by chibenwa** [comment]
test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java [hunk 7]
```diff
@@ -18,44 +18,69 @@
...
 
     @Override
     public int compare(MailboxMessage o1, MailboxMessage o2) {
-        Date date1 = getSentDate(o1);
-        Date date2 = getSentDate(o2);
-        int i = date1.compareTo(date2);
-        
-        // sent date was the same so use the uid as tie-breaker
-        if (i == 0) {
-            return UidComparator.UID.compare(o1, o2);
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java]: removing this does not break anything?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 8]
```diff
@@ -1172,4 +1172,30 @@ public void searchShouldOrderByInternalDateWhenSortOnSentDateAndNoCorrespondingH
                 message1.getUid(),
                 message3.getUid());
     }
+
+    @Test
+    public void searchShouldOrderBySentDateThenInternalInternalDateWhenSortOnSentDateAndNonHomogeneousCorrespondingHeader() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: s/InternalInternal/Internal

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java (hunk 7 above)
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/SentDateComparator.java]: Not at all, as this default behavior is enforced at a higher level.

This line was actually harmful: you could not specify another sort after this one...

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged
